### PR TITLE
Add api function to getOwnPropertyDescriptor and GetOwnPropertyNames

### PR DESCRIPTION
**Pull request by kisbg**
JerryScript-DCO-1.0-Signed-off-by: bence gabor kis kisbg@inf.u-szeged.hu

### CODE REVIEW

File: deps/jerry/v8jerry/v8jerry_value.cpp [hunk 1]
```diff
@@ -81,13 +81,14 @@ JerryValue* JerryValue::GetInternalProperty(JerryValue* key) {
...
-
+    jerry_property_descriptor_t prop_desc;
+    jerry_get_own_property_descriptor (m_value, key, &prop_desc);
+    jerry_value_t descriptor = jerry_from_property_descriptor (&prop_desc);
     return new JerryValue(descriptor);
 }
 
 JerryValue* JerryValue::GetOwnPropertyNames() const {
-    return new JerryValue(JerryIsolate::GetCurrent()->HelperGetOwnPropNames().Call(m_value, NULL, 0));
+    return new JerryValue(jerry_object_get_property_names (m_value, JERRY_PROPERTY_FILTER_TRAVERSE_PROTOTYPE_CHAIN));
```
Comment (zherczeg) [inline deps/jerry/v8jerry/v8jerry_value.cpp]: What should happen with symbols?

**Comment by zherczeg** [comment]
How does this affect test passes?

File: deps/jerry/v8jerry/v8jerry_value.cpp (hunk 1 above)
Comment (rerobika) [reply deps/jerry/v8jerry/v8jerry_value.cpp]: If this function called by `v8::Object::GetOwnPropertyNames` The prototype chain should not be traversed, symbols should be skipped and only enumerable properties should be listed. see (https://github.com/nodejs/node/blob/master/deps/v8/src/api/api.cc#L4446)

**Comment by kisbg** [comment]
@zherczeg  this pull request doesn't change the affect the test passes.

File: deps/jerry/v8jerry/v8jerry_value.cpp (hunk 1 above)
Comment (kisbg) [reply deps/jerry/v8jerry/v8jerry_value.cpp]: i updated the pull request

Comment (rerobika) [reply deps/jerry/v8jerry/v8jerry_value.cpp]: The flags are still not okay.

File: deps/jerry/v8jerry/v8jerry_value.cpp [hunk 2]
```diff
@@ -81,13 +81,14 @@ JerryValue* JerryValue::GetInternalProperty(JerryValue* key) {
 
 JerryValue* JerryValue::GetOwnPropertyDescriptor(const JerryValue& jkey) const {
     jerry_value_t key = jkey.value();
-    jerry_value_t descriptor = JerryIsolate::GetCurrent()->HelperGetOwnPropDesc().Call(m_value, &key, 1);
-
+    jerry_property_descriptor_t prop_desc;
+    jerry_get_own_property_descriptor (m_value, key, &prop_desc);
```
Comment (rerobika) [inline deps/jerry/v8jerry/v8jerry_value.cpp]: what if the result is false?

File: deps/jerry/v8jerry/v8jerry_value.cpp [hunk 3]
```diff
@@ -81,13 +81,23 @@ JerryValue* JerryValue::GetInternalProperty(JerryValue* key) {
...
     return new JerryValue(descriptor);
 }
 
 JerryValue* JerryValue::GetOwnPropertyNames() const {
-    return new JerryValue(JerryIsolate::GetCurrent()->HelperGetOwnPropNames().Call(m_value, NULL, 0));
+    jerry_property_filter_t flags = (jerry_property_filter_t) (JERRY_PROPERTY_FILTER_EXLCUDE_SYMBOLS
+                                                               | JERRY_PROPERTY_FILTER_EXLCUDE_NON_ENUMERABLE);
+    jerry_value_t property_names_val = jerry_object_get_property_names (m_value, flags);
+
+    if(jerry_value_is_array (property_names_val))
```
Comment (rerobika) [inline deps/jerry/v8jerry/v8jerry_value.cpp]: If this condition does not evaluates to true I think an empty local value should be returned in `api.cc`. I think this helper method can be removed.

File: deps/jerry/v8jerry/v8jerry_value.cpp [hunk 4]
```diff
@@ -81,13 +81,23 @@ JerryValue* JerryValue::GetInternalProperty(JerryValue* key) {
 
 JerryValue* JerryValue::GetOwnPropertyDescriptor(const JerryValue& jkey) const {
```
Comment (rerobika) [inline deps/jerry/v8jerry/v8jerry_value.cpp]: Can we remove this helper method?

File: deps/jerry/v8jerry/v8jerry_value.cpp (hunk 2 above)
Comment (rerobika) [reply deps/jerry/v8jerry/v8jerry_value.cpp]: I think an empty local value should be returned in this case.

**Review by rerobika — CHANGES_REQUESTED** [review]

File: deps/jerry/api.cc [hunk 5]
```diff
@@ -1965,8 +1965,19 @@ MaybeLocal<Value> v8::Object::GetOwnPropertyDescriptor(Local<Context> context,
   V8_CALL_TRACE();
   JerryValue* jobject = reinterpret_cast<JerryValue*>(this);
   JerryValue* jkey = reinterpret_cast<JerryValue*>(*key);
+  JerryValue* property_object;
 
-  JerryValue* property_object = jobject->GetOwnPropertyDescriptor(*jkey);
+  jerry_property_descriptor_t prop_desc;
```
Comment (rerobika) [inline deps/jerry/api.cc:1970]: prop_desc should be initialized with  `jerry_init_property_descriptor_fields (&prop_desc);`

File: deps/jerry/api.cc [hunk 6]
```diff
@@ -1965,8 +1965,19 @@ MaybeLocal<Value> v8::Object::GetOwnPropertyDescriptor(Local<Context> context,
...
+
+  if (has_own_property)  {
+      property_object = new JerryValue (jerry_from_property_descriptor (&prop_desc));
+      jerry_free_property_descriptor_fields(&prop_desc);
+      RETURN_HANDLE(Value, context->GetIsolate(), property_object);
+  }
+
+  property_object = new JerryValue ();
+  jerry_free_property_descriptor_fields(&prop_desc);
   RETURN_HANDLE(Value, context->GetIsolate(), property_object);
```
Comment (rerobika) [inline deps/jerry/api.cc:1981]: If the property does not exist you should return an empty handle.
`return MaybeLocal<Value>()`

File: deps/jerry/api.cc [hunk 7]
```diff
@@ -2046,8 +2057,20 @@ MaybeLocal<Array> v8::Object::GetOwnPropertyNames(Local<Context> context) {
...
+  jerry_value_t property_names_val = jerry_object_get_property_names (jobject->value(), flags);
+  JerryValue* names_obj;
+
+  if (jerry_value_is_array (property_names_val))  {
+      names_obj = new JerryValue (property_names_val);
+      jerry_release_value (property_names_val);
+      RETURN_HANDLE(Array, context->GetIsolate(), names_obj);
+  }
+
+  names_obj = new JerryValue (jerry_create_array(0));
```
Comment (rerobika) [inline deps/jerry/api.cc:2071]: Ditto, on error, empty handler should be returned.

File: deps/jerry/api.cc [hunk 8]
```diff
@@ -2046,8 +2057,20 @@ MaybeLocal<Array> v8::Object::GetOwnPropertyNames(Local<Context> context) {
...
+  jerry_release_value (property_names_val);
+  RETURN_HANDLE(Array, context->GetIsolate(), names_obj);
```
Comment (rerobika) [inline deps/jerry/api.cc:2073]:
```c
if (error)
{
return MaybeLocal<Array>();
}

RETURN_HANDLE(...)
```

**Review by rerobika — CHANGES_REQUESTED** [review]

**Pull request closed without merging**